src/depends/NAT/nat.cpp

```diff
@@ -0,0 +1,169 @@
+#include "nat.h"
+#include "libUtils/Logger.h"
+#include <arpa/inet.h> //INET6_ADDRSTRLEN
+#include <cstring>
+#include <iostream>
+#include <miniupnpc/miniupnpc.h>
+#include <miniupnpc/upnpcommands.h>
+
+using namespace std;
+
+NAT::NAT()
+{
+    m_urls = make_shared<UPNPUrls>();
+    m_data = make_shared<IGDdatas>();
+
+    memset(m_urls.get(), 0, sizeof(struct UPNPUrls));
+    memset(m_data.get(), 0, sizeof(struct IGDdatas));
+
+    m_initialized = false;
+
+    struct UPNPDev* devlist;
+    int error = 0;
+
+#if MINIUPNPC_API_VERSION >= 14
+    devlist = upnpDiscover(2000, NULL, NULL, 0, 0, 2, &error);
+#else
+    devlist = upnpDiscover(2000, NULL, NULL, 0, 0, &error);
+#endif
+
+    if (devlist == NULL || error != 0)
+    {
+        LOG_GENERAL(WARNING, "devlist empty or error in discovery");
+        freeUPNPDevlist(devlist);
+        return;
+    }
+    char lan_address[INET6_ADDRSTRLEN];
+
+    int status = UPNP_GetValidIGD(devlist, m_urls.get(), m_data.get(),
+                                  lan_address, sizeof(lan_address));
+
+    if (status != 1)
+    {
+        LOG_GENERAL(WARNING, "Unable to get Valid IGD");
```

Comment:
what is IGD?

Reply:
Internet Gateway Device (eg. router)

Reply:
Maybe to make to friendly to other developer, we can change it to 

`LOG_GENERAL(WARNING, "Unable to get Valid Internet Gateway Device (IGD)");`